parse/src/main/java/com/parse/ParsePlugins.java

```diff
@@ -169,27 +176,170 @@ InstallationId installationId() {
+        if (cacheDir == null) {
+            cacheDir = new File(applicationContext.getCacheDir(), "com.parse");
+        }
+        return createFileDir(cacheDir);
+    }
+
+    /*Migration [START]*/
+
+    /**
+     * Get cache, file object for Parse.
+     *
+     * @param useMigration to migrate old data to new locations before returning cache {@link File} object.
+     */
+    File getCacheDir(Boolean useMigration) {
+        synchronized (lock) {
+            if (useMigration) {
+                runSilentMigration();
+            }
+            return _getCacheDir();
+        }
+    }
+
+    /**
+     * Get files, file object for Parse.
+     *
+     * @param useMigration to migrate old data to new locations before returning cache {@link File} object.
+     */
+    File getFilesDir(Boolean useMigration) {
+        synchronized (lock) {
+            if (useMigration) {
+                runSilentMigration();
             }
-            return createFileDir(filesDir);
+            return _getFilesDir();
         }
     }
+
+    /**
+     * Start old data migrations to new respective locations ({@link #getFilesDir()}, {@link #getCacheDir()}).
+     */
+    public void runSilentMigration() {
+        ArrayList<File> filesToBeMigrated = new ArrayList<>();
+        ParseFileUtils.getAllNestedFiles(
+            getOldParseDir().getAbsolutePath(),
+            filesToBeMigrated
+        );
+        runParseDirFilesMigrationToCacheDir(filesToBeMigrated);
+    }
+
+    public static String replaceLast(String text, String regex, String replacement) {
+        return text.replaceFirst("(?s)" + regex + "(?!.*?" + regex + ")", replacement);
+    }
+
+    /**
+     * Since there is two new different APIs {@link #getFilesDir()}, {@link #getCacheDir()} for caching.
+     * <p></p>
+     * - Config related files, that manages by SDK itself should be migrated to {@link #getFilesDir()}`.
+     * <p></p>
+     * - Others files, that exists in the {@link #getOldParseDir()} should be migrated to {@link #getCacheDir()}.
+     * <p></p>
+     * To identify the config files from {@link #getOldParseDir()},
+     * we are using a hard coded list which defines the SDK config files names.
+     *
+     * @param allFiles Files that needed to be migrated into new location.
+     */
+    private void runParseDirFilesMigrationToCacheDir(ArrayList<File> allFiles) {
+        if (allFiles.isEmpty()) {
+            return;
+        }
+        boolean useFilesDir = false;
+        //Hard coded config file names list.
+        ArrayList<String> configFilesToMigrateToFilesDir = new ArrayList<>();
+        configFilesToMigrateToFilesDir.add("installationId");
+        configFilesToMigrateToFilesDir.add("currentUser");
+        configFilesToMigrateToFilesDir.add("currentConfig");
+        configFilesToMigrateToFilesDir.add("currentInstallation");
+        configFilesToMigrateToFilesDir.add("LocalId");
+        configFilesToMigrateToFilesDir.add("pushState");
```

Comment:
For readability this can be converted to something like:

```java
String[] strArray = {"installationId", "currentUser", "currentConfig", "currentInstallation", ...};
List<String> configFilesToMigrateToFilesDir = Arrays.asList(strArray);
```

Reply:
Thanks. Refactored.